### ISSUE TITLE
[RFC] Evaluate scaffolder templates through dry-run API for testing

### ISSUE THREAD

**Issue by Pike** [backstage/backstage#14280]
**Status:** Open for comments

## Need

Template developers want to test their in-development templates against their installation of backstage.

## Proposal

Have a [CLI script](https://github.com/backstage/backstage/pull/13991) that uses the dry-run API of a backstage installation, and serialize the created workdir to local disk for the developer to test.

With some auth trickery, this could also be used for CI on template repositories.

## Alternatives

* Just rendering a template with nunjucks.
  * This would not utilize all the non-fetch-template parts
* Write a wrapper around the actual dry-run logic to be executed locally.
  * This would require some way to get all custom steps to execute, and would require quite a bit of mocking of backstage infrastructure.

## Risks

This doesn't actually test the survey/questionnaire part of the template.

**Comment by freben** [comment]
Hi,

Just checking, are you aware of the `/create/edit` path that's typically available on your Backstage installation? Also reachable from the top right "kebab menu" in the Create section.

![Screenshot 2022-10-24 at 10 06 51](https://user-images.githubusercontent.com/3097461/197478060-a2cc56f4-460c-4e66-8e13-7377dda7f2d8.png)

Does this correlate with what you need?

**Comment by jhaals** [comment]
Can this be closed with #13991 merged in?

**Comment by Pike** [comment]
> Can this be closed with #13991 merged in?

This issue is to collect feedback on that document, so no.

**Comment by Rugvip** [comment]
Forgot about the relation between things here, got a bit confused by the format I think x) I updated the title to hopefully be a bit more clear about the intent. Perhaps it would be best to update the issue body with references to the code in contrib as well?

**Comment by haho16** [comment]
Hey all,
New to Backstage. Still learning and experimenting with the scaffolder templating. Quick doubt about this:
Is that a standalone tool or does it have a dependency to be run from within a Backstage node? I.e. does it have some system/node dependencies?

**Comment by Pike** [comment]
Yes, it does have node.js dependencies, but it's not required to be run on a machine with backstage installed. To avoid saying node twice for different things. The machine you run it on does need network access to the machine running backstage, though. So it's low on local dependencies, but whether you call that stand-alone or not would be up to you ;-)

**Comment by haho16** [comment]
Hey, @Pike, basically I am having issues with running this on local machine, due to following errors:
```
import { dirname, join, relative } from 'node:path';
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at Object.compileFunction (node:vm:360:18)
    at wrapSafe (node:internal/modules/cjs/loader:1088:15)
    at Module._compile (node:internal/modules/cjs/loader:1123:27)
    at Module._extensions..js (node:internal/modules/cjs/loader:1213:10)
    at Module.load (node:internal/modules/cjs/loader:1037:32)
    at Module._load (node:internal/modules/cjs/loader:878:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:23:47
```
I suspected I am missing a `package.json` file whereby I specify something like:
```
{
  "type": "module"
}
```
and then I get 
```
import type { ScaffolderDryRunResponse } from '@backstage/plugin-scaffolder';
            ^
SyntaxError: Unexpected token '{'
    at Loader.moduleStrategy (internal/modules/esm/translators.js:145:18)
    at async link (internal/modules/esm/module_job.js:64:21)
```
I know I am doing something wrong or missing some npm dependencies, but not sure which?

I am on `Node.js v18.12.1`. I am running it in the following way:
```sh
node scaffolder-dry.js http://remotehost/ <template-directory> template.yml <output-directory>
```
Excuse my ignorance if it is something obvious :)

**Comment by brandonphan** [comment]
Would also like some clarification on how to use the script, not super familiar with typescript and couldn't get it working while messing around with it. My apologies if something is obvious.

**Comment by Pike** [comment]
Let's keep this open.

**Comment by jessebye** [comment]
For a bit more context, with Cookiecutter we had a simple way to test templates locally: just run `cookiecutter` and examine the output. This was critical because when building new templates, you need to iterate rapidly and don't want to add any extra steps or dependencies like running the template through an active Backstage instance.

We also used Cookiecutter in CI pipelines to build some basic tests. At least, making sure the template compiled and didn't have any errors.

With the switch to nunjucks, we don't have any good way to test locally *or* in CI. This is a big step backward for us in terms of functionality.

**Comment by Gladdstone** [comment]
Could really use some more details on how this is meant to be run. Not clear on why this is a .md. As a .ts the path to the package.json may have changed since this was created, but the package.json it appears to be referencing is not configured properly to execute this as a module. Not currently able to get this to function.

**Comment by rmartine-ias** [comment]
@Gladdstone I got it working, here's how I modified it: https://gist.github.com/rmartine-ias/4fe26f27b6d520806bd26fa0f1607fa2

If this is missing anything, let me know. It's not super pretty, but it runs.

`yarn install`, and run with
`npx ts-node --esm src/index.ts http://localhost:7007/ path/to/template/folder values.yml out --token "$FRONTEND_TOKEN"` (also kind of hacky, but it runs)

I got `FRONTEND_TOKEN` by going to backstage in my browser, and inspecting the request headers to authenticated resources (e.g. asking about entities) -- copy the JWT after "Bearer: " and then `export FRONTEND_TOKEN="whatever"`. `values.yml` should include all required values for the parameters step.

**Comment by electrofelix** [comment]
I've been bumping into an issue recently using the template editor directly, it appears that when using `replace(' ', '-')` to set up a value, trying to click Create in the preview to run through the scaffolder dry-run results in a 403 error. Switching this to `replace(' ', '\-')`, gets through fine and it appears that using the escaping has no ill effects when done for real.

The unescaped form works fine when using templates for real, it only appears to cause an issue when sent to the dry-run endpoint. I'm guessing a serialisation issue perhaps? Using roadie and they confirmed they weren't seeing anything in their logs regarding the forbidden error.

Happy to report this as a separate issue, just figured some people might end up i this thread.

Btw, having this is fantastic, so I'm hopeful that work will continue and build out a CLI tool to support along the lines of the scripts referenced above.

**Comment by freben** [comment]
@electrofelix Could you make a separate issue with some reproduction steps for that?

**Comment by sp71** [comment]
Any chance we can increase the priority for this CLI to be official? It's an amazing tool

**Comment by benjdlambert** [comment]
@sp71 yep it's something that's on our radar, I think we probably want to wait for some of #15999 before to settle before taking on some of the CLI work. Thinks like authentication etc need to be settled for CLI's first.

**Comment by rmartine-ias** [comment]
not stale, still want this

**Comment by Rugvip** [comment]
@Pike what do you think about moving this to an official `@backstage/plugin-scaffolder-cli` package? (👀 @benjdlambert)

**Comment by Pike** [comment]
naggitynagnag

**Comment by Pike** [comment]
naggitynagnag

**Comment by benjdlambert** [comment]
@Pike that aught to keep the stalebot away. Still blocked on the auth issue at the minute though, which we should hopefully get some time in Q1 to start taking a look at.

**Comment by cmoulliard** [comment]
Any update about the integration of the dry-run part of the backstage client ?

**Comment by dakr0013** [comment]
> > The new backend makes this easier. I have a less-hacky version locally, but I haven't been able to actually run through it in earnest due to day job things.
> 
> Can you share it please ?

In case you or someone else still need it. Here is how I have done it:

Add this to catalog-info.yaml
```yaml
backend:
  auth:
    externalAccess:
      - type: static
        options:
          token: ${STATIC_TOKEN_CICD}
          subject: cicd-scaffolder-dry-run
        accessRestrictions:
          - plugin: scaffolder
```
I can use value of `STATIC_TOKEN_CICD` in my CI pipeline (which does a scaffolder dry-run) to auth against backstage.

See https://backstage.io/docs/auth/service-to-service-auth/#static-tokens
